File: components/gameBoard/src/main/java/yi/component/board/edits/RemoveNodeEdit.java
package yi.component.board.edits;

import yi.core.go.GameModel;
import yi.core.go.GameNode;

import java.util.Objects;

public final class RemoveNodeEdit extends UndoableEdit {

    private final GameNode parentOfNodeToRemove;
    private final GameNode nodeToRemove;

    public RemoveNodeEdit(GameNode nodeToRemove) {
        if (nodeToRemove.isRoot()) {
            throw new IllegalArgumentException("Cannot remove root node");
        }

        this.nodeToRemove = Objects.requireNonNull(nodeToRemove);
        this.parentOfNodeToRemove =  Objects.requireNonNull(nodeToRemove.getParent());
    }

    @Override
    protected boolean _rollbackEdit(GameModel gameModel) {
        if (!parentOfNodeToRemove.isContinuationOf(gameModel.getRootNode())) {
            throw new IllegalStateException("Attempting to re-do but parent node is no longer part of the game tree");
        }

        gameModel.submitNode(parentOfNodeToRemove, nodeToRemove);
        return true;
    }

    @Override
    protected boolean _performEdit(GameModel gameModel) {
        gameModel.removeNodeSubtree(nodeToRemove);
        return true;
    }

    @Override
    protected boolean canRollback() {
        return false;
    }
}

File: yi.core/src/main/kotlin/yi/core/go/editor/edit/RemoveNodeEdit.kt
package yi.core.go.editor.edit

import yi.core.go.GameModel
import yi.core.go.GameNode

class RemoveNodeEdit constructor(private val nodeToRemove: GameNode) : GameModelEdit {

    private var parentOfNodeToRemove: GameNode? = null

    init {
        check(!nodeToRemove.isRoot()) { "Cannot remove root node" }
    }

    override fun rollbackChanges(model: GameModel) {
        check (parentOfNodeToRemove!!.isContinuationOf(model.getRootNode())) {
            "Attempting to undo but parent is no longer part of the game tree"
        }
        model.editor.addNode(parentOfNodeToRemove!!, nodeToRemove)
    }

    override fun performChanges(model: GameModel): Boolean {
        check(nodeToRemove.isContinuationOf(model.getRootNode())) {
            "Attempting to remove node but the node is no longer part of the game tree"
        }
        parentOfNodeToRemove = nodeToRemove.parent!!
        model.editor.removeNode(nodeToRemove)
        return true
    }

    override fun dispose() {
        nodeToRemove.dispose()
    }
}